cookbooks/tomography_based_plate_motions/plugins/tomography_based_plate_motions.cc

```diff
@@ -0,0 +1,1726 @@
+
+            Assert(unscaled_viscosity_outputs != nullptr,ExcInternalError());
+
+            for (unsigned int q=0; q<output.size(); ++q)
+              output[q] = unscaled_viscosity_outputs->output_values[0][q];
+          }
+      };
+    }
+  }
+
+  namespace MaterialModel
+  {
+    template <int dim>
+    void
+    TomographyBasedPlateMotions<dim>::initialize()
+    {
+      // Get reference viscosity profile from the ascii data
+      if (use_depth_dependent_viscosity)
+        reference_viscosity_coordinates = reference_viscosity_profile->get_interpolation_point_coordinates();
+
+      // Get column index for density scaling
+      if (use_depth_dependent_rho_vs)
+        {
+          rho_vs_depth_profile.initialize(this->get_mpi_communicator());
+          density_scaling_index = rho_vs_depth_profile.get_column_index_from_name("density_scaling");
+        }
+
+      if (use_depth_dependent_thermal_expansivity)
+        {
+          thermal_expansivity_profile.initialize(this->get_mpi_communicator());
+          thermal_expansivity_column_index = thermal_expansivity_profile.get_column_index_from_name("thermal_expansivity");
+        }
+
+      if (use_depth_dependent_dT_vs)
+        {
+          dT_vs_depth_profile.initialize(this->get_mpi_communicator());
+          temperature_scaling_index = dT_vs_depth_profile.get_column_index_from_name("temperature_scaling");
+        }
+
+      // Get column for crustal depths
+      std::set<types::boundary_id> surface_boundary_set;
+      surface_boundary_set.insert(this->get_geometry_model().translate_symbolic_boundary_name_to_id("top"));
+      crustal_boundary_depth.initialize(surface_boundary_set, 1);
+
+      // The update() function updates the profile that stores the laterally averaged viscosity.
+      // This is needed to compute the viscosity in the material model (since the viscosities are rescaled
+      // so that the lateral average matches the reference profile). Since the viscosity depends on both
+      // temperature and velocity, we want to update the lateral average profile after each temperature
+      // and Stokes solve.
+      this->get_signals().post_stokes_solver.connect([&](const SimulatorAccess<dim> &,
+                                                         const unsigned int ,
+                                                         const unsigned int ,
+                                                         const SolverControl &,
+                                                         const SolverControl &)
+      {
+        this->update();
+      });
+
+      this->get_signals().post_advection_solver.connect([&](const SimulatorAccess<dim> &,
+                                                            const unsigned int ,
+                                                            const unsigned int ,
+                                                            const SolverControl &)
+      {
+        this->update();
+      });
```

Comment:
If I understand it right, your model is instantaneous and so you never actually solve for the temperature. In that case, the comment is not quite right, nor is it necessary to connect to the `post_advection_solver`.

This is mostly for clarification to myself: You *could* update the comment and remove the second `connect()` call. But perhaps it is best to leave it in place anyway just in case someone wanted to extend our model to the time-dependent case?